android/PhysicalWeb/app/src/main/java/org/physical_web/physicalweb/MainActivity.java

```diff
@@ -83,28 +80,91 @@ public boolean onOptionsItemSelected(MenuItem item) {
+    checkingPermissions = true;
+    if (!bluetoothAdapter.isEnabled()) {
+      Intent enableBtIntent = new Intent(BluetoothAdapter.ACTION_REQUEST_ENABLE);
+      startActivityForResult(enableBtIntent, REQUEST_ENABLE_BT);
+      return;
+    }
+    ensureLocationPermissionIsEnabled();
+  }
+
+  @Override
+  protected void onActivityResult (int requestCode, int resultCode, Intent data) {
+    Log.d(TAG, "onActivityResult");
+    if (requestCode == REQUEST_ENABLE_BT && resultCode == -1) {
+      ensureLocationPermissionIsEnabled();
+      return;
+    }
+    Toast.makeText(this, getString(R.string.bt_on), Toast.LENGTH_LONG).show();
+    finish();
+  }
+
   @Override
   protected void onResume() {
     super.onResume();
-    Log.d(TAG, "resumed MainActivity");
-    BluetoothManager btManager = (BluetoothManager) getSystemService(BLUETOOTH_SERVICE);
-    BluetoothAdapter btAdapter = btManager != null ? btManager.getAdapter() : null;
-    if (btAdapter == null) {
-      Toast.makeText(getApplicationContext(),
-              R.string.error_bluetooth_support, Toast.LENGTH_LONG).show();
-      finish();
-      return;
+    if(!checkingPermissions){
+      Log.d(TAG, "resumed MainActivity");
+      BluetoothManager btManager = (BluetoothManager) getSystemService(BLUETOOTH_SERVICE);
+      BluetoothAdapter btAdapter = btManager != null ? btManager.getAdapter() : null;
+      if (btAdapter == null) {
+        Toast.makeText(getApplicationContext(),
+            R.string.error_bluetooth_support, Toast.LENGTH_LONG).show();
+        finish();
+        return;
+      }
+      if (checkIfUserHasOptedIn()) {
+        ensureBluetoothIsEnabled(btAdapter);
+      } else {
+        // Show the oob activity
+        Intent intent = new Intent(this, OobActivity.class);
+        startActivity(intent);
+      }
     }
-    if (checkIfUserHasOptedIn()) {
-      ensureBluetoothIsEnabled(btAdapter);
-      showNearbyBeaconsFragment();
-      Intent intent = new Intent(this, ScreenListenerService.class);
-      startService(intent);
-    } else {
-      // Show the oob activity
-      Intent intent = new Intent(this, OobActivity.class);
-      startActivity(intent);
+    
+  }
+
+  private void finishLoad() {
+    showNearbyBeaconsFragment();
+    Intent intent = new Intent(this, ScreenListenerService.class);
+    startService(intent);
+  }
+
+  private void ensureLocationPermissionIsEnabled() {
+    if (Build.VERSION.SDK_INT >=23 && ContextCompat.checkSelfPermission(getApplicationContext(),android.Manifest.
+        permission.ACCESS_COARSE_LOCATION) != PackageManager.PERMISSION_GRANTED) {
+      ActivityCompat.requestPermissions(this,
+          new String[]{android.Manifest.permission.ACCESS_COARSE_LOCATION},REQUEST_LOCATION);
+      return;
     }
+    checkingPermissions = false;
+    finishLoad();
+  }
+
+  @Override
+  public void onRequestPermissionsResult(int requestCode,
+          String permissions[], int[] grantResults) {
+      switch (requestCode) {
```

Comment:
function body, indent 2 spaces

Also, switch statements should indent the body 2 spaces, and each case should indent another 2 spaces, eg:

switch (val) {
__case 0:
____// do stuff
____break;
}

https://google.github.io/styleguide/javaguide.html#s4.8.4-switch